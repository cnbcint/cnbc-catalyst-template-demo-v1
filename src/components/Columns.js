import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { breakpoints } from "../helpers/Layout";
const Container = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
  flex-direction: column;
  @media (min-width: ${breakpoints.mobile}) {
    flex-direction: row;
  }
  > div {
    width: calc(100% - 0.5rem);
    @media (min-width: ${breakpoints.mobile}) {
      width: calc(${(props) => 100 / props.count}% - 1rem);
    }
    box-sizing: border-box;
  }
`;
/**
 *
 * This component will distribute any children in columns the amount of columns is determined
 * by the amouunt of children nested.
 */
export default function Columns(props) {
  return (
    <Container
      count={props.children.length}
      className={props.className}
      style={props.style}
    >
      {props.children.map((item, i) => {
        return (
          <div
            key={i}
            className={
              item.props.className ? item.props.className + "__wrapper" : ""
            }
          >
            {item}
          </div>
        );
      })}
    </Container>
  );
}

Columns.propTypes = {
  className: PropTypes.string,
  children: PropTypes.array.isRequired,
  /**
   * Style object in React format
   */
  style: PropTypes.object,
};
