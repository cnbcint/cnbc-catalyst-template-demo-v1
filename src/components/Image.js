import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { breakpoints } from "../helpers/Layout";
import useAnimation from "../hooks/useAnimation";
const ImageElement = styled.img`
  width: 100%;
`;
/**
Responsive image component with lazyload tag. 
This components required images for desktop tablet and mobile screensizes. 
Each image can have it's own aspect ratio to achive more flexibility on design.
The image will take 100% of space available from parent container.
 */
export default function Image(props) {
  const { ref: refImage } = useAnimation({
    animationName: "fadeIn",
    duration: 500,
    delay: 0,
    loop: true,
    autoStart: true,
  });

  return (
    <picture className={props.className}>
      <source
        media={`(min-width:${breakpoints.tablet})`}
        srcSet={props.source.desktop}
      />
      <source
        media={`(min-width:${breakpoints.mobile})`}
        srcSet={props.source.tablet}
      />
      <ImageElement
        ref={refImage}
        src={props.source.mobile}
        style={props.style}
        alt={props.alt}
        loading="lazy"
      />
    </picture>
  );
}

Image.propTypes = {
  className: PropTypes.string,
  alt: PropTypes.string.isRequired,
  /**
   Image url source for desktop tablet and mobile
   */
  source: PropTypes.shape({
    mobile: PropTypes.string,
    tablet: PropTypes.string,
    desktop: PropTypes.string,
  }).isRequired,
  /**
   * React CSS overrides to IMG element
   */
  style: PropTypes.object,
};

Image.defaultProps = {
  className: "",
};
