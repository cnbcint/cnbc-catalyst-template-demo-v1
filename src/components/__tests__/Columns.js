import React from "react";
import Columns from "../Columns";
import Image from "../Image";
import { render } from "@testing-library/react";
import "jest-styled-components";

beforeAll(() => {
  window.IntersectionObserver = class {
    observe() {
      return null;
    }
    disconnect() {}
  };
});

describe("Columns", () => {
  it("Renders correctly", () => {
    const { container } = render(
      <Columns>
        <Image
          source={{
            desktop:
              "http://localhost:8080/images/responsive/output/image05-tablet.jpg",
            tablet:
              "http://localhost:8080/images/responsive/output/image05-tablet.jpg",
            mobile:
              "http://localhost:8080/images/responsive/output/image05-mobile.jpg",
          }}
          alt="Alt Text"
        />
        <Image
          source={{
            desktop:
              "http://localhost:8080/images/responsive/output/image06-tablet.jpg",
            tablet:
              "http://localhost:8080/images/responsive/output/image06-tablet.jpg",
            mobile:
              "http://localhost:8080/images/responsive/output/image06-mobile.jpg",
          }}
          alt="Alt Text"
        />
      </Columns>
    );
    expect(container).toMatchSnapshot();
  });

  it("Renders correctly with 2 columns", () => {
    const { container } = render(
      <Columns>
        <Image
          source={{
            desktop:
              "http://localhost:8080/images/responsive/output/image05-tablet.jpg",
            tablet:
              "http://localhost:8080/images/responsive/output/image05-tablet.jpg",
            mobile:
              "http://localhost:8080/images/responsive/output/image05-mobile.jpg",
          }}
          alt="Alt Text"
        />
        <Image
          source={{
            desktop:
              "http://localhost:8080/images/responsive/output/image06-tablet.jpg",
            tablet:
              "http://localhost:8080/images/responsive/output/image06-tablet.jpg",
            mobile:
              "http://localhost:8080/images/responsive/output/image06-mobile.jpg",
          }}
          alt="Alt Text"
        />
      </Columns>
    );
    expect(container.getElementsByTagName("img").length).toBe(2);
  });
});

it("Has flex css property", () => {
  const { container } = render(
    <Columns>
      <Image
        source={{
          desktop:
            "http://localhost:8080/images/responsive/output/image05-tablet.jpg",
          tablet:
            "http://localhost:8080/images/responsive/output/image05-tablet.jpg",
          mobile:
            "http://localhost:8080/images/responsive/output/image05-mobile.jpg",
        }}
        alt="Alt Text"
      />
      <Image
        source={{
          desktop:
            "http://localhost:8080/images/responsive/output/image06-tablet.jpg",
          tablet:
            "http://localhost:8080/images/responsive/output/image06-tablet.jpg",
          mobile:
            "http://localhost:8080/images/responsive/output/image06-mobile.jpg",
        }}
        alt="Alt Text"
      />
    </Columns>
  );
  expect(container.firstChild).toHaveStyleRule("display", "flex");
});

it("Has flex-direction column css property", () => {
  const { container } = render(
    <Columns>
      <Image
        source={{
          desktop:
            "http://localhost:8080/images/responsive/output/image05-tablet.jpg",
          tablet:
            "http://localhost:8080/images/responsive/output/image05-tablet.jpg",
          mobile:
            "http://localhost:8080/images/responsive/output/image05-mobile.jpg",
        }}
        alt="Alt Text"
      />
      <Image
        source={{
          desktop:
            "http://localhost:8080/images/responsive/output/image06-tablet.jpg",
          tablet:
            "http://localhost:8080/images/responsive/output/image06-tablet.jpg",
          mobile:
            "http://localhost:8080/images/responsive/output/image06-mobile.jpg",
        }}
        alt="Alt Text"
      />
    </Columns>
  );
  expect(container.firstChild).toHaveStyleRule("flex-direction", "column");
});

afterAll(() => {
  window.IntersectionObserver = null;
});
