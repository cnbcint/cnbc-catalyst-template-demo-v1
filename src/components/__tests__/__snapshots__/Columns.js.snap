// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Columns Renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 > div {
  width: calc(100% - 0.5rem);
  box-sizing: border-box;
}

.c1 {
  width: 100%;
}

@media (min-width:768px) {
  .c0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:768px) {
  .c0 > div {
    width: calc(50% - 1rem);
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class=""
    >
      <picture
        class=""
      >
        <source
          media="(min-width:1280px)"
          srcset="http://localhost:8080/images/responsive/output/image05-tablet.jpg"
        />
        <source
          media="(min-width:768px)"
          srcset="http://localhost:8080/images/responsive/output/image05-tablet.jpg"
        />
        <img
          alt="Alt Text"
          class="c1 pre-animated"
          loading="lazy"
          src="http://localhost:8080/images/responsive/output/image05-mobile.jpg"
        />
      </picture>
    </div>
    <div
      class=""
    >
      <picture
        class=""
      >
        <source
          media="(min-width:1280px)"
          srcset="http://localhost:8080/images/responsive/output/image06-tablet.jpg"
        />
        <source
          media="(min-width:768px)"
          srcset="http://localhost:8080/images/responsive/output/image06-tablet.jpg"
        />
        <img
          alt="Alt Text"
          class="c1 pre-animated"
          loading="lazy"
          src="http://localhost:8080/images/responsive/output/image06-mobile.jpg"
        />
      </picture>
    </div>
  </div>
</div>
`;
